File: PreviousQuestions/217.cpp
#include <iostream>
#include <cstring>
using namespace std;
int p;
struct martix
{
    long long data[3][3];
    martix() {memset(data,0,sizeof(data));}
}ans,base;
martix operator *(martix res,martix temp) {
        martix cur;
        for (int i=1; i<=2; i++)
            for (int j=1; j<=2; j++)
                for (int k=1; k<=2; k++)
                    cur.data[i][j]= (cur.data[i][j]%p+(res.data[i][k]%p)*(temp.data[k][j]%p)%p)% p;
        return cur;
}
long long n,m;
void power(long long n)
{
    while (n){
        if (n&1) ans=base*ans;
        base=base*base;
        n>>=1;
    }
}
int f(int n)
{
  if (n<=2) {return ans.data[n][1];}
    else {
        power(n-2);
        return ans.data[1][1];
    }
}
int main()
{
    base.data[1][1]=1,base.data[1][2]=1,base.data[2][1]=1,base.data[2][2]=0;
    ans.data[1][1]=1,ans.data[2][1]=1;
    cin>>n>>m>>p;
    cout<<((f(n+2)-1)%f(m))%p;
    return 0;
}

File: PreviousQuestions/217.py
     
def f(n)->int:
    base=[[1,1],[1,0]]
    ans=[[1,0],[1,0]]
    if n<3:
        return ans[n-1][0]
    else:
        n-=2
        while n:
            if n&1:
                ans= [[sum(map(lambda a: a[0]*a[1], zip(l, s))) for l in zip(*ans)] for s in base]
            base= [[sum(map(lambda a: a[0]*a[1], zip(l, s))) for l in zip(*base)] for s in base]
            n>>=1
        return ans[0][0]


n,m,p=map(int,input().split())
print(((f(n+2)-1) % f(m)) %p)